### PR TITLE
GDPopt update

### PR DESCRIPTION
**Pull request by qtothec**
## Summary/Motivation:
Various enhancements and updates

## Changes proposed in this PR:
- Move integer to binary reformulation away from GDPopt
- Allow solution of MINLP subproblems
- Simplification of component lists (for maintaining consistency across clones)
- tighter integer cuts for disjunctive subproblems
- Improvement of timing reporting
- Improved nonlinear objective handling
- Include citation guide
- Move to date-based versioning
- Minor utility function edits to support MindtPy development

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by qtothec** [comment]
Coverage here is pretty terrible because we are missing the ability to test with MC++. Looks better when run locally.
coverage

**Comment by ghackebeil** [comment]
Is this something I can try to install on the Travis test images? If so, send me a bash script that downloads and compiles it, and I’ll incorporate it into the docker images we use.

**Comment by qtothec** [comment]
Thanks, Gabe. Unfortunately, I don't currently have a clean way to build MC++ and create the shared object link to the interface that @johnnybatesjr wrote, although that's on the docket of items that I plan to work with @jsiirola on next week.

**Comment by jsiirola** [comment]
I am just about done with a standard set of tools (and another lightweight Registry) for handling downloading (binary) files. I expect that we will get that integrated this week, too.

File: pyomo/contrib/gdpopt/mip_solve.py [hunk 1]
```diff
@@ -68,14 +66,27 @@ def solve_linear_GDP(linear_GDP_model, solve_data, config):
         # resolve with a solver option flag on.
         results, terminate_cond = distinguish_mip_infeasible_or_unbounded(
             m, config)
+    if terminate_cond is tc.unbounded:
+        # Solution is unbounded. Add an arbitrary bound to the objective and resolve.
+        obj_bound = 1E15
```
Comment (qtothec) [inline pyomo/contrib/gdpopt/mip_solve.py]: Add comment here to indicate how this comes about.

**Pull request merged**